modules/xml/filterx-parse-xml.c

```diff
@@ -97,50 +97,60 @@ _set_error(GError **error, const gchar *format, ...)
   va_end(va);
 }
 
-
-typedef struct XmlElemContext_
-{
-  FilterXObject *current_obj;
-  FilterXObject *parent_obj;
-} XmlElemContext;
-
-static void
-_elem_context_set_current_obj(XmlElemContext *self, FilterXObject *current_obj)
+void
+xml_elem_context_set_current_obj(XmlElemContext *self, FilterXObject *current_obj)
 {
   filterx_object_unref(self->current_obj);
   self->current_obj = filterx_object_ref(current_obj);
 }
 
-static void
-_elem_context_set_parent_obj(XmlElemContext *self, FilterXObject *parent_obj)
+void
+xml_elem_context_set_parent_obj(XmlElemContext *self, FilterXObject *parent_obj)
 {
   filterx_object_unref(self->parent_obj);
   self->parent_obj = filterx_object_ref(parent_obj);
 }
 
-static void
-_elem_context_free(XmlElemContext *self)
+void
+xml_elem_context_free(XmlElemContext *self)
 {
-  _elem_context_set_current_obj(self, NULL);
-  _elem_context_set_parent_obj(self, NULL);
+  xml_elem_context_set_current_obj(self, NULL);
+  xml_elem_context_set_parent_obj(self, NULL);
   g_free(self);
 }
 
-static XmlElemContext *
-_elem_context_new(FilterXObject *parent_obj, FilterXObject *current_obj)
+XmlElemContext *
+xml_elem_context_new(FilterXObject *parent_obj, FilterXObject *current_obj)
 {
   XmlElemContext *self = g_new0(XmlElemContext, 1);
-  _elem_context_set_parent_obj(self, parent_obj);
-  _elem_context_set_current_obj(self, current_obj);
+  xml_elem_context_set_parent_obj(self, parent_obj);
+  xml_elem_context_set_current_obj(self, current_obj);
   return self;
 }
 
 
-typedef struct FilterXGeneratorFunctionParseXml_
+void
+filterx_parse_xml_user_data_init_instance(FilterXParseXmlUserData *self)
 {
-  FilterXGeneratorFunction super;
-  FilterXExpr *xml_expr;
-} FilterXGeneratorFunctionParseXml;
+  self->xml_elem_context_stack = g_queue_new();
+  self->free_fn = filterx_parse_xml_user_data_free_method;
+}
+
+void
+filterx_parse_xml_user_data_free_method(FilterXParseXmlUserData *self)
+{
+  g_queue_free_full(self->xml_elem_context_stack, (GDestroyNotify) xml_elem_context_free);
+  g_free(self);
```

Comment:
I usually have the g_free() call in the wrapper that calls the free_fn() method, so the free_fn() can be optional.

Reply:
Thanks, fixed.